core/lib/zksync_core/src/native_token_fetcher/mod.rs

```diff
@@ -84,20 +89,35 @@ impl NativeTokenFetcher {
     }
 
     pub(crate) async fn run(&self, stop_receiver: watch::Receiver<bool>) -> anyhow::Result<()> {
+        let mut network_consecutive_errors = 0;
         loop {
             if *stop_receiver.borrow() {
                 tracing::info!("Stop signal received, native_token_fetcher is shutting down");
                 break;
             }
 
-            let conversion_rate = reqwest::get(format!("{}/conversion_rate", &self.config.host))
-                .await?
-                .json::<u64>()
-                .await
-                .unwrap();
-
-            self.latest_to_eth_conversion_rate
-                .store(conversion_rate, std::sync::atomic::Ordering::Relaxed);
+            match reqwest::get(format!("{}/conversion_rate", &self.config.host)).await {
+                Ok(response) => {
+                    let conversion_rate = response.json::<u64>().await?;
+                    self.latest_to_eth_conversion_rate
+                        .store(conversion_rate, std::sync::atomic::Ordering::Relaxed);
+                    network_consecutive_errors = 0;
+                }
+                Err(err) => {
+                    network_consecutive_errors += 1;
+
+                    tracing::error!(
+                        "Failed to fetch native token conversion rate from the server: {err}"
+                    );
+
+                    if network_consecutive_errors >= MAX_CONSECUTIVE_NETWORK_ERRORS {
+                        vlog::capture_message(&err.to_string(), vlog::AlertLevel::Warning);
+
+                        // reset the error counter to prevent sending multiple sentry errors consecutively
+                        network_consecutive_errors = 0;
```

Comment:
Nice! However, if the polling interval is set to 3 seconds, we would still get 2 alerts per minute, which is probably not a good idea.
Additionally, the retry logic seems to be mixed with the actual application logic.
Maybe we declare a helper structure that would encapsulate it. Something like this (pseudocode/not prettified):

```rust
struct ErrorReporter {
  current_try: u64,
  alert_spawned: bool,
}

impl ErrorReporter {
  const MAX_CONSECUTIVE_NETWORK_ERRORS: u8 = 10;
  fn reset(&mut self) {
    self.current_try = 0;
    self.alert_spawned = false;
  }
  fn process(&mut self, err: Error) {
    self.current_try += std::cmp::min(self.current_try + 1, Self::MAX_CONSECUTIVE_NETWORK_ERRORS);  
    tracing::error!("Failed to fetch native token conversion rate from the server: {err}");
    if self.current_try >= Self::MAX_CONSECUTIVE_NETWORK_ERRORS && !self.alert_spawned {
        vlog::capture_message(&err.to_string(), vlog::AlertLevel::Warning);
        self.alert_spawned = true;
    }
  }
}
```

Reply:
This way we will only spawn a single alert per server going down. We will still, however, spawn multiple alerts if the server goes up and down, but that's probably a good thing: otherwise someone may check that in this exact moment "huh, it's up right now, probably it's fine".

Reply:
I agree, moving the error handling to a separate structure keeps code clean. I have two questions about your pseudocode:

1. In what case should we `reset()` the `ErrorReporter`? 
2. Should we keep this struct inside the same file? Or move it so it can be ussed by multiple components?

Reply:
I've implemented a simple `ErrorReporter`, be sure to check it out whenever you can.